src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,201 @@
+/*
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AsyncTransferFacilitator.h"
+
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace bdx {
+
+AsyncTransferFacilitator::~AsyncTransferFacilitator()
+{
+    Reset();
+}
+
+void AsyncTransferFacilitator::Reset()
+{
+    mExchange.Release();
+    mTransfer.Reset();
+}
+
+CHIP_ERROR AsyncTransferFacilitator::SendMessage(TransferSession::OutputEvent & event)
+{
+
+    VerifyOrReturnError(mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    Messaging::SendFlags sendFlags;
+
+    // All messages sent from the Sender expect a response, except for a StatusReport which would indicate an error and
+    // the end of the transfer.
+    if (!event.msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kExpectResponse);
+    }
+
+    auto & msgTypeData = event.msgTypeData;
+
+    Messaging::ExchangeContext * ec = mExchange.Get();
+    VerifyOrReturnError(ec != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    // Set the response timeout on the exchange before sending the message.
+    ec->SetResponseTimeout(mTimeout);
+
+    CHIP_ERROR err = ec->SendMessage(msgTypeData.ProtocolId, msgTypeData.MessageType, std::move(event.MsgData), sendFlags);
+
+    // If there's an error sending the message, release the exchange.
+    if (err != CHIP_NO_ERROR)
+    {
+        mExchange.Release();
+    }
+    return err;
+}
+
+CHIP_ERROR AsyncTransferFacilitator::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                                       System::PacketBufferHandle && payload)
+{
+
+    VerifyOrReturnError(mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+        mTransfer.AbortTransfer(AsyncResponder::GetBdxStatusCodeFromChipError(err));
```

Comment:
So this will handle creating whatever events need to happen to clean things up, I assume?  If so, that should be documented.

Reply:
yes it ll generate a status report from the BDXTransferSession which when we receive will clean up. i will document this.

Reply:
i added a comment. Please let me know if that's enough